File: pkg/common/socks5.go
package common

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"strconv"
)

// SOCKS5 protocol constants defined by RFC 1928 and RFC 1929. They are shared
// between the inbound frontend (pkg/socksproxy) and the outbound chaining
// backend (pkg/adapter/socksproxy).
const (
	// SOCKS5Version is the SOCKS protocol version byte.
	SOCKS5Version byte = 0x05

	// Method selection codes used during the initial negotiation.
	SOCKS5MethodNoAuth           byte = 0x00
	SOCKS5MethodUsernamePassword byte = 0x02
	SOCKS5MethodNoAcceptable     byte = 0xFF

	// SOCKS5AuthVersion is the version byte of the RFC 1929 username/password
	// sub-negotiation.
	SOCKS5AuthVersion byte = 0x01

	// SOCKS5CmdConnect identifies the CONNECT command. BIND and UDP ASSOCIATE
	// are not supported by puppy.
	SOCKS5CmdConnect byte = 0x01

	// Address-type codes for DST.ADDR / BND.ADDR.
	SOCKS5AtypIPv4   byte = 0x01
	SOCKS5AtypDomain byte = 0x03
	SOCKS5AtypIPv6   byte = 0x04

	// Reply (REP) codes returned in SOCKS5 replies.
	SOCKS5RepSuccess              byte = 0x00
	SOCKS5RepGeneralFailure       byte = 0x01
	SOCKS5RepConnectionNotAllowed byte = 0x02
	SOCKS5RepNetworkUnreachable   byte = 0x03
	SOCKS5RepHostUnreachable      byte = 0x04
	SOCKS5RepConnectionRefused    byte = 0x05
	SOCKS5RepTTLExpired           byte = 0x06
	SOCKS5RepCmdNotSupported      byte = 0x07
	SOCKS5RepAddrTypeNotSupported byte = 0x08
)

// SOCKS5ReplyText maps a SOCKS5 reply code to a human-readable description.
func SOCKS5ReplyText(rep byte) string {
	switch rep {
	case SOCKS5RepSuccess:
		return "succeeded"
	case SOCKS5RepGeneralFailure:
		return "general SOCKS server failure"
	case SOCKS5RepConnectionNotAllowed:
		return "connection not allowed by ruleset"
	case SOCKS5RepNetworkUnreachable:
		return "network unreachable"
	case SOCKS5RepHostUnreachable:
		return "host unreachable"
	case SOCKS5RepConnectionRefused:
		return "connection refused"
	case SOCKS5RepTTLExpired:
		return "TTL expired"
	case SOCKS5RepCmdNotSupported:
		return "command not supported"
	case SOCKS5RepAddrTypeNotSupported:
		return "address type not supported"
	default:
		return "unknown error (0x" + strconv.FormatUint(uint64(rep), 16) + ")"
	}
}

// ReadSOCKS5Address reads a SOCKS5 DST.ADDR (or BND.ADDR) followed by the
// 2-byte big-endian DST.PORT from r, decoding IPv4, IPv6, and domain address
// types. The returned host is the literal IP (stringified) or domain name; the
// returned port is the decoded port number. An error is returned for unknown
// address types or short reads.
func ReadSOCKS5Address(r io.Reader, atyp byte) (host string, port uint16, err error) {
	switch atyp {
	case SOCKS5AtypIPv4:
		var addr [4]byte
		if _, err := io.ReadFull(r, addr[:]); err != nil {
			return "", 0, fmt.Errorf("read IPv4 address: %w", err)
		}
		host = net.IP(addr[:]).String()
	case SOCKS5AtypIPv6:
		var addr [16]byte
		if _, err := io.ReadFull(r, addr[:]); err != nil {
			return "", 0, fmt.Errorf("read IPv6 address: %w", err)
		}
		host = net.IP(addr[:]).String()
	case SOCKS5AtypDomain:
		var lenByte [1]byte
		if _, err := io.ReadFull(r, lenByte[:]); err != nil {
			return "", 0, fmt.Errorf("read domain length: %w", err)
		}
		domain := make([]byte, lenByte[0])
		if _, err := io.ReadFull(r, domain); err != nil {
			return "", 0, fmt.Errorf("read domain: %w", err)
		}
		host = string(domain)
	default:
		return "", 0, fmt.Errorf("unknown address type 0x%02x", atyp)
	}

	var portBytes [2]byte
	if _, err := io.ReadFull(r, portBytes[:]); err != nil {
		return "", 0, fmt.Errorf("read port: %w", err)
	}
	port = binary.BigEndian.Uint16(portBytes[:])
	return host, port, nil
}

File: rust/crates/puppy-core/src/socks5.rs
//! SOCKS5 protocol constants and address reader.

use std::io;
use std::net::{Ipv4Addr, Ipv6Addr};

use tokio::io::{AsyncRead, AsyncReadExt};

/// SOCKS5 protocol version.
pub const VERSION: u8 = 0x05;

/// Method: No authentication required.
pub const METHOD_NO_AUTH: u8 = 0x00;
/// Method: Username/password (RFC 1929).
pub const METHOD_USERNAME_PASSWORD: u8 = 0x02;
/// Method: No acceptable methods.
pub const METHOD_NO_ACCEPTABLE: u8 = 0xFF;

/// RFC 1929 auth version.
pub const AUTH_VERSION: u8 = 0x01;

/// Command: CONNECT.
pub const CMD_CONNECT: u8 = 0x01;

/// Address type: IPv4.
pub const ATYP_IPV4: u8 = 0x01;
/// Address type: Domain name.
pub const ATYP_DOMAIN: u8 = 0x03;
/// Address type: IPv6.
pub const ATYP_IPV6: u8 = 0x04;

/// Reply code: succeeded.
pub const REP_SUCCESS: u8 = 0x00;
/// Reply code: general SOCKS server failure.
pub const REP_GENERAL_FAILURE: u8 = 0x01;
/// Reply code: connection not allowed by ruleset.
pub const REP_CONNECTION_NOT_ALLOWED: u8 = 0x02;
/// Reply code: network unreachable.
pub const REP_NETWORK_UNREACHABLE: u8 = 0x03;
/// Reply code: host unreachable.
pub const REP_HOST_UNREACHABLE: u8 = 0x04;
/// Reply code: connection refused.
pub const REP_CONNECTION_REFUSED: u8 = 0x05;
/// Reply code: TTL expired.
pub const REP_TTL_EXPIRED: u8 = 0x06;
/// Reply code: command not supported.
pub const REP_CMD_NOT_SUPPORTED: u8 = 0x07;
/// Reply code: address type not supported.
pub const REP_ADDR_TYPE_NOT_SUPPORTED: u8 = 0x08;

/// Returns the RFC 1928 text for a reply code.
///
/// Unknown codes produce `unknown error (0x<hex>)` matching `strconv.FormatUint(.., 16)`.
pub fn reply_text(rep: u8) -> String {
	match rep {
		REP_SUCCESS => "succeeded".to_string(),
		REP_GENERAL_FAILURE => "general SOCKS server failure".to_string(),
		REP_CONNECTION_NOT_ALLOWED => "connection not allowed by ruleset".to_string(),
		REP_NETWORK_UNREACHABLE => "network unreachable".to_string(),
		REP_HOST_UNREACHABLE => "host unreachable".to_string(),
		REP_CONNECTION_REFUSED => "connection refused".to_string(),
		REP_TTL_EXPIRED => "TTL expired".to_string(),
		REP_CMD_NOT_SUPPORTED => "command not supported".to_string(),
		REP_ADDR_TYPE_NOT_SUPPORTED => "address type not supported".to_string(),
		other => format!("unknown error (0x{other:x})"),
	}
}

/// Reads DST.ADDR + DST.PORT from a SOCKS5 request/reply given `atyp`.
///
/// The underlying `io::Error` kind is preserved (so `ErrorKind::UnexpectedEof`
/// propagates like `errors.Is(err, io.EOF)`).
pub async fn read_socks5_address<R: AsyncRead + Unpin>(
	reader: &mut R,
	atyp: u8,
) -> io::Result<(String, u16)> {
	let host = match atyp {
		ATYP_IPV4 => {
			let mut buf = [0u8; 4];
			read_full(reader, &mut buf, "read IPv4 address").await?;
			Ipv4Addr::from(buf).to_string()
		}
		ATYP_IPV6 => {
			let mut buf = [0u8; 16];
			read_full(reader, &mut buf, "read IPv6 address").await?;
			Ipv6Addr::from(buf).to_string()
		}
		ATYP_DOMAIN => {
			let mut len_buf = [0u8; 1];
			read_full(reader, &mut len_buf, "read domain length").await?;
			let len = len_buf[0] as usize;
			let mut buf = vec![0u8; len];
			read_full(reader, &mut buf, "read domain").await?;
			String::from_utf8(buf).map_err(|e| {
				io::Error::new(io::ErrorKind::InvalidData, format!("read domain: {e}"))
			})?
		}
		other => {
			return Err(io::Error::other(format!(
				"unknown address type 0x{other:02x}"
			)))
		}
	};

	let mut port_buf = [0u8; 2];
	read_full(reader, &mut port_buf, "read port").await?;
	let port = u16::from_be_bytes(port_buf);

	Ok((host, port))
}

/// Wraps `read_exact` so the underlying error kind is preserved and the message
/// is prefixed with `what`, matching `fmt.Errorf("%s: %w", what, err)`.
async fn read_full<R: AsyncRead + Unpin>(
	reader: &mut R,
	buf: &mut [u8],
	what: &str,
) -> io::Result<()> {
	reader
		.read_exact(buf)
		.await
		.map_err(|e| io::Error::new(e.kind(), format!("{what}: {e}")))?;
	Ok(())
}
